gnomad/utils/constraint.py

```diff
@@ -13,6 +14,170 @@
 logger.setLevel(logging.INFO)
 
 
+def annotate_with_mu(
+    ht: hl.Table,
+    mutation_ht: hl.Table,
+    mu_annotation: str = "mu_snp",
+) -> hl.Table:
+    """
+    Annotate SNP mutation rate for the input Table.
+
+    .. note::
+        `ht` is expected to include annotations that `mutation_ht` is keyed by, but these annotations don't need to be the keys of `ht`.
+
+    :param ht: Input Table to annotate.
+    :param mutation_ht: Mutation rate Table.
+    :param mu_annotation: The name of mutation rate annotation in `mutation_ht`. Defaults to 'mu_snp'.
+    :return: Table with mutational rate annotation added (default name for annotation is 'mu_snp').
+    """
+    mu = mutation_ht.index(*[ht[k] for k in mutation_ht.key])[mu_annotation]
+    return ht.annotate(
+        **{mu_annotation: hl.case().when(hl.is_defined(mu), mu).or_error("Missing mu")}
+    )
+
+
+def count_variants_by_group(
+    ht: hl.Table,
+    freq_expr: hl.expr.ArrayExpression = None,
+    freq_meta_expr: hl.expr.ArrayExpression = None,
+    count_singletons: bool = False,
+    count_downsamplings: Optional[Tuple[str]] = (),
+    additional_grouping: Optional[Tuple[str]] = (),
+    partition_hint: int = 100,
+    omit_methylation: bool = False,
+    use_table_group_by: bool = False,
+    singleton_expr: hl.expr.BooleanExpression = None,
+    max_af: Optional[float] = None,
+) -> Union[hl.Table, Any]:
+    """
+    Count number of observed or possible variants by context, ref, alt, and optionally methylation_level.
+
+    Function counts variants in `freq_expr` based on specified criteria (singleton or in downsamplings or None), and aggregates variant counts based on
+    'context', 'ref', 'alt', 'methylation_level' (optional), and `additional_grouping` provided.
+
+    If `freq_expr` is not given and variants in downsamplings or singleton variants need to be counted, `freq_expr` defaults to `ht.freq`. If variants in downsamplings
+    needs to be counted and `freq_meta_expr` is None, `freq_meta_expr` defaults to `ht.freq_meta`. If variants in singleton needs to be counted and singleton_expr is None,
+    `singleton_expr` defaults to `freq_expr[0].AC == 1`.
+
+    Function will return a Table with annotations used for grouping ('context', 'ref', 'alt', 'methylation_level' (optional), `additional_grouping`) and 'variant_count' annotation.
+
+    .. note::
+        The following annotations should be present in `ht`:
+        - ref - the middle base of `context`
+        - alt - the alternate base
+        - context - trinucleotide genomic context
+        - methylation_level - methylation level
+        - freq - Allele frequency information (AC, AN, AF, homozygote count) in gnomAD release.
```

Comment:
This line and the one after it have periods, but the ones before it don't. I think either is fine, but just change for consistency